### PR TITLE
ci: add developer-tools-manylinux2014 stack

### PR DESCRIPTION
**Pull request by eugeneswalker**
Add Developer Tools CI stack built using `manylinux2014` container image.

@tgamblin @tldahlgren

### CODE REVIEW

**Review by tldahlgren — DISMISSED** [review]
LGTM

File: share/spack/gitlab/cloud_pipelines/stacks/developer-tools-manylinux2014/spack.yaml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      operating_system: centos7
+      target: x86_64
+      modules: []
+      environment: {}
+      extra_rpaths: []
+
+  ci:
+    pipeline-gen:
+    - build-job:
+        image: ecpe4s/manylinux2014:2024.03.11
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/developer-tools-manylinux2014/spack.yaml:86]: I think having this in ghcr would be better, but not required for a first pass on this.

File: share/spack/gitlab/cloud_pipelines/stacks/developer-tools-manylinux2014/spack.yaml [hunk 2]
```diff
@@ -0,0 +1,89 @@
+spack:
+  view: false
+  packages:
+    all:
+      require: target=x86_64_v3
+  concretizer:
+    unify: true
+  definitions:
+  - default_specs:
+      # editors
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/developer-tools-manylinux2014/spack.yaml:10]: Does it make sense to include clang and gnu compilers here? Or do you include these in the build cache create by building the image?

**Review by kwryankrattiger — APPROVED** [review]
Just some minor comments, they don't need to be addressed in this PR, I think the getting this stack started will be useful to a lot of people.

File: share/spack/gitlab/cloud_pipelines/stacks/developer-tools-manylinux2014/spack.yaml (hunk 2 above)
Comment (trws) [reply share/spack/gitlab/cloud_pipelines/stacks/developer-tools-manylinux2014/spack.yaml:10]: Both gcc and clang are built as part of building this stack, but the gcc is pulled in by emacs. The clang is explicitly built by the llvm item. Neither are really meant to be used _as compilers_ where I use them currently, they're there to provide clang-tools* and libgccjit and similar tooling.  There's no reason they can't be, but they wont necessarily be base-system compatible if they come out of a bin-cache.

File: share/spack/gitlab/cloud_pipelines/stacks/developer-tools-manylinux2014/spack.yaml (hunk 1 above)
Comment (eugeneswalker) [reply share/spack/gitlab/cloud_pipelines/stacks/developer-tools-manylinux2014/spack.yaml:86]: I can't push to the ghcr Spack registry -- can you mirror this image over there and I can update the reference?

**Comment by eugeneswalker** [comment]
Using @tgamblin's PR to look more closely at the concretizer decisions
* https://github.com/spack/spack/pull/41150

Here is the output of `script --return concretize-debug-nondefaults.log --command "spack -e . solve -d"`
* [concretize-debug-nondefaults.log](https://github.com/spack/spack/files/14652321/concretize-debug-nondefaults.log) (file includes color codes

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**